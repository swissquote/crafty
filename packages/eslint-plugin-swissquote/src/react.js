const { addMissingRules } = require("./utils");

// EcmaScript 6 specific configuration
module.exports = {
  settings: {
    "import/resolver": {
      node: {
        extensions: [".js", ".jsx", ".json"]
      }
    },
    react: {
      version: "detect"
    }
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true
    }
  },
  rules: {
    // Recommended rules override
    "@swissquote/swissquote/react/display-name": 0, // Disabled as it generates false positives
    "@swissquote/swissquote/react/jsx-no-duplicate-props": [
      "error",
      { ignoreCase: true }
    ],
    "@swissquote/swissquote/react/no-deprecated": "warn",
    "@swissquote/swissquote/react/prop-types": [
      "warn",
      { skipUndeclared: true }
    ],

    // Swissquote Rules
    // Disabled for now as it has false positives on inline handlers
    // TODO :: re-enable jsx-handler-names once https://github.com/yannickcr/eslint-plugin-react/issues/2832 is fixed
    "@swissquote/swissquote/react/jsx-handler-names": "off",
    "@swissquote/swissquote/react/jsx-pascal-case": "error",
    "@swissquote/swissquote/react/no-did-mount-set-state": "error",
    "@swissquote/swissquote/react/no-did-update-set-state": "error",
    "@swissquote/swissquote/react/no-redundant-should-component-update":
      "error",
    "@swissquote/swissquote/react/no-typos": "error",
    "@swissquote/swissquote/react/no-unused-state": "error",
    "@swissquote/swissquote/react/no-will-update-set-state": "error",
    "@swissquote/swissquote/react/prefer-es6-class": ["error", "always"],
    "@swissquote/swissquote/react/prefer-stateless-function": [
      "error",
      { ignorePureComponents: true }
    ],
    "@swissquote/swissquote/react/void-dom-elements-no-children": "error"
  }
};

// Add all recommended configurations from eslint-plugin-react
addMissingRules(
  require("eslint-plugin-react").configs.recommended.rules,
  module.exports.rules
);

// Add all recommended configurations from eslint-plugin-react-hooks
addMissingRules(
  require("eslint-plugin-react-hooks").configs.recommended.rules,
  module.exports.rules
);

// Disable all the rules from eslint-plugin-react that are handled by Prettier
addMissingRules(
  require("eslint-config-prettier/react").rules,
  module.exports.rules
);
