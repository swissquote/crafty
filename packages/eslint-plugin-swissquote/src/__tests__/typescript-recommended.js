/* global describe, it, expect */

const { prepareCLIEngine, lint } = require("../../test_utils");
const engine = prepareCLIEngine("recommended");

it("Warns on console.log", () => {
  const result = lint(engine, `
const foo = window;

// Use a TypeScript 3.7 feature to make sure it works
if (foo?.bar?.baz) {
  console.log(foo.bar.baz);
}
`, "file.ts");

  expect(result.messages).toMatchSnapshot();
  expect(result.warningCount).toBe(1);
  expect(result.errorCount).toBe(0);
});


it("Uses sonar plugin", () => {
  const result = lint(
    engine,
    `
/* global openWindow, closeWindow, moveWindowToTheBackground */

function changeWindow(param: number) {
  if (param === 1) {
    openWindow();
  } else if (param === 2) {
    closeWindow();
  } else if (param === 1) {
    // Noncompliant    ^
    moveWindowToTheBackground();
  }
}
`, "file.ts"
  );

  expect(result.messages).toMatchSnapshot();
  expect(result.warningCount).toBe(0);
  expect(result.errorCount).toBe(2);
});

it("Works with complex types", () => {
  const result = lint(
    engine,
    `
import * as React from "react";

type Writeable<T> = { -readonly [P in keyof T]-?: T[P] };

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

function assignRef<T>(ref: T, setter: React.Ref<T>) {
  if (!setter || !ref) {
    return;
  }

  if (typeof setter === "function") {
    setter(ref);
    return;
  }

  if (setter.hasOwnProperty("current")) {
    (setter as Writeable<React.RefObject<T>>).current = ref;
    return;
  }

  throw new Error(
    "Your ref must be a function or an object created with React.createRef"
  );
}

export type triggerFunction = (options: {
  isOpen: boolean;
  caret: React.ReactNode;
  props: Record<string, unknown>;
}) => React.ReactNode;

export interface ButtonGroupProps extends React.HTMLAttributes<HTMLDivElement> {
  innerRef?: React.Ref<HTMLDivElement>;
}

export interface SplitButtonProps extends Omit<ButtonGroupProps, "innerRef"> {
  other: React.ReactNode;
  open?: boolean;
  trigger: React.ReactNode | triggerFunction;

  /**
   * Ref to get the component
   */
  innerRef?: React.Ref<HTMLElement>;
}

export interface SplitButtonState {
  open?: boolean;
}

export default class SplitButton extends React.Component<
  SplitButtonProps,
  SplitButtonState
> {
  constructor(props) {
    super(props);

    this.state = {
      open: props.open || false
    };
  }

  private target: Element;

  ref = node => {
    this.target = getElement(node);
    assignRef(this.target, this.props.innerRef);
  };

  render(trigger) {
    const { trigger } = this.props;
    if (typeof trigger !== "function") {
      // eslint-disable-next-line @swissquote/swissquote/react/jsx-handler-names
      return <button innerRef={this.ref}>{trigger}</button>;
    }

    return (trigger as triggerFunction)({
      props: { innerRef: this.ref },
      isOpen: this.state.open,
      caret: <span className={this.props.classes("Caret")} />
    });
  }
}
`,
    "Component.tsx"
  );

  expect(result.messages).toMatchSnapshot();
  expect(result.warningCount).toBe(0);
  expect(result.errorCount).toBe(0);
});