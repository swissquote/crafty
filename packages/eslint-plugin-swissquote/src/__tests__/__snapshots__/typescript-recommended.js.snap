// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Uses sonar plugin 1`] = `
Array [
  Object {
    "column": 10,
    "endColumn": 22,
    "endLine": 3,
    "line": 3,
    "message": "'changeWindow' is defined but never used.",
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "@swissquote/swissquote/@typescript-eslint/no-unused-vars",
    "severity": 2,
  },
  Object {
    "column": 14,
    "endColumn": 25,
    "endLine": 8,
    "line": 8,
    "message": "This branch duplicates the one on line 4",
    "nodeType": "BinaryExpression",
    "ruleId": "@swissquote/swissquote/sonarjs/no-identical-conditions",
    "severity": 2,
  },
]
`;

exports[`Warns on console.log 1`] = `
Array [
  Object {
    "column": 3,
    "endColumn": 14,
    "endLine": 5,
    "line": 5,
    "message": "Unexpected console statement.",
    "messageId": "unexpected",
    "nodeType": "MemberExpression",
    "ruleId": "no-console",
    "severity": 1,
  },
]
`;

exports[`Works with complex types 1`] = `Array []`;
